src/indexing/binding_builder.rs

```diff
@@ -1,54 +1,51 @@
-use std::{fmt, marker::PhantomData};
+use std::fmt;
 
 use crate::{utils::UniqueStack, HashSet, IndexMap};
 
-use super::{BindVariableError, IndexKey};
+use super::BindVariableError;
 
-pub(super) struct BindingBuilder<K, V, M> {
+pub(super) struct BindingBuilder<M: IndexMap> {
     bindings: M,
     /// A stack of the missing keys.
     ///
     /// It's important that we process the keys first-in last-out, to ensure
     /// that we process the dependencies we added to the stack before processing
     /// the same key again.
-    missing_keys: UniqueStack<K>,
-    exclude_keys: HashSet<K>,
-    _phantom: PhantomData<V>,
+    missing_keys: UniqueStack<M::Key>,
+    exclude_keys: HashSet<M::Key>,
 }
 
-impl<K: Clone, V, M: Clone> Clone for BindingBuilder<K, V, M> {
+impl<M: IndexMap> Clone for BindingBuilder<M> {
```

Comment:
Note you can skip this now and just `#[derive(Clone)]`

Reply:
oh good catch, thanks!